src/storage/write_ahead_log/log_manager.cpp

```diff
@@ -12,50 +12,118 @@ void LogManager::Process() {
       buffer = flush_queue_.front();
       flush_queue_.pop();
     }
-    for (LogRecord &record : IterableBufferSegment<LogRecord>(buffer)) {
-      if (record.RecordType() == LogRecordType::COMMIT) {
-        auto *commit_record = record.GetUnderlyingRecordBodyAs<CommitRecord>();
-
-        // If a transaction is read-only, then the only record it generates is its commit record. This commit record is
-        // necessary for the transaction's callback function to be invoked, but there is no need to serialize it, as
-        // it corresponds to a transaction with nothing to redo.
-        if (!commit_record->IsReadOnly()) SerializeRecord(record);
-        commits_in_buffer_.emplace_back(commit_record->Callback(), commit_record->CallbackArg());
-        // Not safe to mark read only transactions as the transactions are deallocated preemptively without waiting for
-        // logging (there is nothing to log after all)
-        if (!commit_record->IsReadOnly()) commit_record->Txn()->log_processed_ = true;
-      } else {
-        // Any record that is not a commit record is always serialized.`
-        SerializeRecord(record);
-      }
-    }
+    // Serialize the Redo buffer and release it to the buffer pool
+    IterableBufferSegment<LogRecord> task_buffer(buffer);
+    SerializeTaskBuffer(&task_buffer);
     buffer_pool_->Release(buffer);
   }
+  // Mark the last buffer that was written to as full
+  if (buffer_to_write_ != nullptr) {
+    MarkBufferFull();
+  }
   Flush();
 }
 
 void LogManager::Flush() {
-  out_.Persist();
+  // Set the flag for the log writer thread to persist the buffers to disk
+  do_persist_ = true;
+  // Wait for the log writer thread to persist the logs
+  while (do_persist_) {
+  }
+  // Execute the callbacks for the transactions that have been persisted
   for (auto &callback : commits_in_buffer_) callback.first(callback.second);
   commits_in_buffer_.clear();
 }
 
+void LogManager::SerializeTaskBuffer(IterableBufferSegment<LogRecord> *const task_buffer) {
```

Comment:
You need some comments explaining the various names you are using (e.g. task buffer)